### PR TITLE
feat(calendar): Eva 1.3

### PR DESCRIPTION
**Pull request by yggg**
### Please read and mark the following check list before creating a pull request:

 - [x] I read and followed the [CONTRIBUTING.md](https://github.com/akveo/nebular/blob/master/CONTRIBUTING.md) guide.
 - [x] I read and followed the [New Feature Checklist](https://github.com/akveo/nebular/blob/master/DEV_DOCS.md#new-feature-checklist) guide.
 
 #### Short description of what this resolves:
Breaking Changes:
  - `NbCalendarHeaderComponent` removed. Along with that `showHeader` property removed from following components:
    - `NbBaseCalendarComponent`
    - `NbCalendarRangeComponent`
    - `NbCalendarComponent`
    - `NbBasePicker`
    - `NbBasePickerComponent`

  - `medium` and `size-medium` css classes were removed along with `medium`/`isMedium` properties from the following classes:
    - `NbCalendarDayPickerComponent`
    - `NbCalendarMonthPickerComponent`
    - `NbCalendarYearPickerComponent`
    - `NbBaseCalendarComponent`
    - `NbCalendarWeekNumberComponent`

  - Properties `date`, `changeMode`, `isRtl` removed from `NbCalendarPageableNavigationComponent`

  - File `nb-calendar-week-number.component` renamed to `calendar-week-number.component`

  - `YEARS_IN_VIEW, YEARS_IN_ROW` constants removed. Use `NbCalendarYearModelService`s `getYearsInView` and `getRowInView` methods.

  - `NbCalendarNavigationComponent` replaced with `NbCalendarViewModeComponent`.

  - `initYears` method removed from `NbCalendarYearPickerComponent`.

  - `calendar-range-cells.ts` file split into `calendar-range-day-cell.component.ts`, `calendar-range-month-cell.component.ts` and `calendar-range-year-cell.component.ts`.

  - Following theme properties were removed:
    - `calendar-body-height`
    - `calendar-header-text-color`
    - `calendar-header-text-font-family`
    - `calendar-header-title-text-font-size`
    - `calendar-header-title-text-font-weight`
    - `calendar-header-title-text-line-height`
    - `calendar-header-sub-title-text-font-size`
    - `calendar-header-sub-title-text-font-weight`
    - `calendar-header-sub-title-text-line-height`
    - `calendar-navigation-button-width`
    - `calendar-large-body-height`
    - `calendar-cell-in-range-background-color`
    - `calendar-cell-disabled-background-color`
    - `datepicker-text-color`
    - `datepicker-arrow-size`

### CODE REVIEW

File: src/framework/theme/components/calendar-kit/components/calendar-week-number/calendar-week-number.component.ts [hunk 1]
```diff
@@ -11,10 +11,12 @@ import { NbCalendarSize } from '../../model';
 @Component({
   selector: 'nb-calendar-week-numbers',
   template: `
-    <div class="sign">{{ weekNumberSymbol }}</div>
-    <div class="week-cell" *ngFor="let week of getWeeks()">{{ week }}</div>
+    <div class="sign-container">
+      <div class="sign">{{ weekNumberSymbol }}</div>
+    </div>
+    <div class="week-number" *ngFor="let week of getWeeks()">{{ week }}</div>
```
Comment (Tibing) [inline src/framework/theme/components/calendar-kit/components/calendar-week-number/calendar-week-number.component.ts]: Since you're already doing enhancements, why not to rewrite `getWeeks()` functionality? As I see it, `getWeeks()` result depends only on `weeks` input, so, don't you think it would be better to recalculate `getWeeks()` result in `weeks` setter and then just query it?

File: src/framework/theme/components/calendar-kit/components/calendar-week-number/calendar-week-number.component.ts [hunk 2]
```diff
@@ -11,10 +11,12 @@ import { NbCalendarSize } from '../../model';
 @Component({
   selector: 'nb-calendar-week-numbers',
   template: `
-    <div class="sign">{{ weekNumberSymbol }}</div>
-    <div class="week-cell" *ngFor="let week of getWeeks()">{{ week }}</div>
+    <div class="sign-container">
```
Comment (Tibing) [inline src/framework/theme/components/calendar-kit/components/calendar-week-number/calendar-week-number.component.ts]: Do we really need an additional container here? Are you sure we can't style it properly without it?

File: src/framework/theme/components/calendar-kit/components/calendar-year-picker/calendar-year-cell.component.spec.ts [hunk 3]
```diff
@@ -42,6 +42,8 @@ describe('Component: NbCalendarYearCell', () => {
   });
 
   it('should contain cell class', () => {
+    component.date = new Date();
+    fixture.detectChanges();
```
Comment (Tibing) [inline src/framework/theme/components/calendar-kit/components/calendar-year-picker/calendar-year-cell.component.spec.ts]: Why do we need it? How did it work before?)

File: src/framework/theme/components/calendar-kit/components/calendar-year-picker/calendar-year-cell.component.ts [hunk 4]
```diff
@@ -19,9 +19,12 @@ import { NbCalendarCell } from '../../model';
 
 @Component({
   selector: 'nb-calendar-year-cell',
-  template: `{{ year }}`,
+  template: `
+    <div class="cell-content">
```
Comment (Tibing) [inline src/framework/theme/components/calendar-kit/components/calendar-year-picker/calendar-year-cell.component.ts]: I already added a comment, but again, do we really need this additional element?

File: src/framework/theme/components/calendar-kit/components/calendar-year-picker/calendar-year-picker.component.ts [hunk 5]
```diff
@@ -62,37 +60,23 @@ export class NbCalendarYearPickerComponent<D> implements OnChanges {
...
+  @HostBinding('class.size-large')
   get large() {
     return this.size === NbCalendarSize.LARGE;
   }
 
   years: D[][];
 
-  constructor(protected dateService: NbDateService<D>) {
-  }
+  constructor(
```
Comment (Tibing) [inline src/framework/theme/components/calendar-kit/components/calendar-year-picker/calendar-year-picker.component.ts]: I'm not sure, but I believe we always format constructors another way:
```
  constructor(protected dateService: NbDateService<D>,
              protected yearModelService: NbCalendarYearModelService<D>) {
  }
```
BTW, I'm pretty sure it fits in one line

File: src/framework/theme/components/calendar-kit/services/calendar-year-model.service.spec.ts [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+  let yearsInView: number;
+  let yearsInRow: number;
+
+  beforeEach(() => {
+    yearModelService = new NbCalendarYearModelService(new NbNativeDateService('en-US'));
+    yearsInView = yearModelService.getYearsInView();
+    yearsInRow = yearModelService.getYearsInRow();
+  });
+
+  it('should return correct years range', () => {
```
Comment (Tibing) [inline src/framework/theme/components/calendar-kit/services/calendar-year-model.service.spec.ts]: Would you mind to add more tests to cover corner cases? I'm dead sure that one test for such functionality is not enough

File: src/framework/theme/components/calendar/_calendar.component.theme.scss [hunk 7]
```diff
@@ -8,94 +8,60 @@
...
 
-    &.in-range {
-      background: nb-theme(calendar-cell-in-range-background-color);
-
-      &:not(.today, :hover) {
-        border-radius: 0;
+    &:not(.has-navigation) {
+      nb-calendar-week-numbers,
+      nb-calendar-days-names {
+        border-top: 0;
```
Comment (Tibing) [inline src/framework/theme/components/calendar/_calendar.component.theme.scss]: Don't you think it ought to be configured thought the theme?

File: src/framework/theme/components/calendar/base-calendar.component.ts [hunk 8]
```diff
@@ -72,9 +72,11 @@ export class NbBaseCalendarComponent<D, T> implements OnInit {
   @Input() visibleDate: D;
 
   /**
-   * Determines should we show calendars header or not.
+   * Determines should we show calendars navigation or not.
```
Comment (Tibing) [inline src/framework/theme/components/calendar/base-calendar.component.ts]:
```suggestion
   * Determines whether we should show calendar navigation or not.
```
I believe it will sound more native

File: src/framework/theme/components/calendar/base-calendar.component.ts [hunk 9]
```diff
@@ -142,24 +141,61 @@ export class NbBaseCalendarComponent<D, T> implements OnInit {
...
+  }
+
+  navigatePrev() {
+    switch (this.activeViewMode) {
+      case NbCalendarViewMode.DATE:
+        return this.prevMonth();
+      case NbCalendarViewMode.MONTH:
+        return this.prevYear();
+      case NbCalendarViewMode.YEAR:
+        return this.prevYears();
```
Comment (Tibing) [inline src/framework/theme/components/calendar/base-calendar.component.ts]: Default case?

File: src/framework/theme/components/calendar/base-calendar.component.ts [hunk 10]
```diff
@@ -142,24 +141,61 @@ export class NbBaseCalendarComponent<D, T> implements OnInit {
...
+  }
+
+  navigateNext() {
+    switch (this.activeViewMode) {
+      case NbCalendarViewMode.DATE:
+        return this.nextMonth();
+      case NbCalendarViewMode.MONTH:
+        return this.nextYear();
+      case NbCalendarViewMode.YEAR:
+        return this.nextYears();
```
Comment (Tibing) [inline src/framework/theme/components/calendar/base-calendar.component.ts]: Default case?

File: src/framework/theme/components/calendar/calendar-range-cells.ts [hunk 11]
```diff
@@ -155,19 +216,149 @@ export class NbCalendarRangeYearCellComponent<D> implements NbCalendarCell<D, Nb
...
+@Component({
+  selector: 'nb-calendar-range-month-cell',
+  template: `
+    <div class="cell-content">
+      {{ month }}
+    </div>
+  `,
+  changeDetection: ChangeDetectionStrategy.OnPush,
+})
+export class NbCalendarRangeMonthCellComponent<D> extends NbBaseCalendarRangeCell<D>
```
Comment (Tibing) [inline src/framework/theme/components/calendar/calendar-range-cells.ts]: Would you mind splitting this file? I believe we need to move each component to a separate file because this file has too many lines.

File: src/framework/theme/components/calendar-kit/components/calendar-day-picker/calendar-day-cell.component.ts [hunk 12]
```diff
@@ -63,6 +66,9 @@ export class NbCalendarDayCellComponent<D> implements NbCalendarCell<D, D> {
     return this.smallerThanMin() || this.greaterThanMax() || this.dontFitFilter();
   }
 
+  @HostBinding('class.day-cell')
```
[the remark this answers is not in the document]
Comment (yggg) [reply src/framework/theme/components/calendar-kit/components/calendar-day-picker/calendar-day-cell.component.ts]: Inheritance

File: src/framework/theme/components/calendar-kit/components/calendar-navigation/calendar-view-mode.component.spec.ts [hunk 13]
```diff
@@ -6,31 +6,28 @@
...
+    fixture = TestBed.createComponent<NbCalendarViewModeComponent<Date>>(NbCalendarViewModeComponent);
     component = fixture.componentInstance;
     componentEl = fixture.debugElement.nativeElement;
   });
 
   it('should render date', () => {
     component.date = new Date(2018, 6, 23);
     fixture.detectChanges();
-    expect(componentEl.querySelector('button').textContent).toContain('Jul 2018');
+    expect(componentEl.querySelector('button').textContent).toContain('July 2018');
```
[the remark this answers is not in the document]
Comment (yggg) [reply src/framework/theme/components/calendar-kit/components/calendar-navigation/calendar-view-mode.component.spec.ts]: Calendar shows full month name now

[the remark this answers is not in the document]
Comment (Tibing) [reply src/framework/theme/components/calendar-kit/components/calendar-navigation/calendar-view-mode.component.spec.ts]: Ok, will it be listed as a breaking change at the release?

File: src/framework/theme/components/calendar-kit/components/calendar-navigation/calendar-view-mode.component.ts [hunk 14]
```diff
@@ -0,0 +1,72 @@
...
+    protected dateService: NbDateService<D>,
+    protected yearModelService: NbCalendarYearModelService<D>,
+  ) {}
+
+  getText(): string {
+    if (!this.date) {
+      return '';
+    }
+
+    switch (this.viewMode) {
```
[the remark this answers is not in the document]
Comment (yggg) [reply src/framework/theme/components/calendar-kit/components/calendar-navigation/calendar-view-mode.component.ts]: > we don't have validation for this.viewMode
We have type set on `viewMode`

> default value for that switch/case statement
Looks like silent fail. Why it's better?

[the remark this answers is not in the document]
Comment (Tibing) [reply src/framework/theme/components/calendar-kit/components/calendar-navigation/calendar-view-mode.component.ts]: I agree with you, partially. If we don't have a default value, then, we ought to fire an exception or something, otherwise, we still have silent fail)

File: src/framework/theme/components/calendar-kit/components/calendar-year-picker/calendar-year-cell.component.spec.ts (hunk 3 above)
Comment (yggg) [reply src/framework/theme/components/calendar-kit/components/calendar-year-picker/calendar-year-cell.component.spec.ts]: It worked because `detectChanges` was never called -> no view rendered, just a host element.

File: src/framework/theme/components/calendar-kit/components/calendar-navigation/calendar-pageable-navigation.component.scss [hunk 15]
```diff
@@ -9,31 +9,11 @@
...
-    @include nb-ltr() {
-      margin-left: auto;
-      margin-right: 0.4rem;
-    }
-    @include nb-rtl() {
-      margin-left: 0.4rem;
-      margin-right: auto;
-    }
+  nb-calendar-view-mode {
+    margin: 0 0.5rem;
```
[the remark this answers is not in the document]
Comment (yggg) [reply src/framework/theme/components/calendar-kit/components/calendar-navigation/calendar-pageable-navigation.component.scss]: Removed completely as view mode component was removed from pageable nav.
835f2a02f8c8fb84850b5a88a1c856da19b6fa6e

File: src/framework/theme/components/calendar-kit/components/calendar-day-picker/_calendar-day-picker.component.theme.scss [hunk 16]
```diff
@@ -0,0 +1,15 @@
+/**
+ * @license
+ * Copyright Akveo. All Rights Reserved.
+ * Licensed under the MIT License. See License.txt in the project root for license information.
+ */
+
+@mixin nb-calendar-day-picker-theme {
+  nb-calendar-day-picker {
```
[the remark this answers is not in the document]
Comment (yggg) [reply src/framework/theme/components/calendar-kit/components/calendar-day-picker/_calendar-day-picker.component.theme.scss]: Done bc1a64741986b39529eac78bdc4eb5a530b08c87

File: src/framework/theme/components/calendar-kit/components/calendar-days-names/calendar-days-names.component.ts [hunk 17]
```diff
@@ -4,8 +4,8 @@
  * Licensed under the MIT License. See License.txt in the project root for license information.
  */
 
-import { ChangeDetectionStrategy, Component, OnInit } from '@angular/core';
-import { NbCalendarDay } from '../../model';
+import { ChangeDetectionStrategy, Component, OnInit, Input, HostBinding } from '@angular/core';
```
[the remark this answers is not in the document]
Comment (yggg) [reply src/framework/theme/components/calendar-kit/components/calendar-days-names/calendar-days-names.component.ts]: b14c9c74344878e893bc6861e73d863707735340

File: src/framework/theme/components/calendar-kit/components/calendar-month-picker/calendar-month-cell.component.ts [hunk 18]
```diff
@@ -62,19 +68,26 @@ export class NbCalendarMonthCellComponent<D> implements NbCalendarCell<D, D> {
...
+  protected monthStart(): D {
     return this.dateService.getMonthStart(this.date);
   }
 
-  private monthEnd(): D {
+  protected monthEnd(): D {
     return this.dateService.getMonthEnd(this.date);
   }
+
+  protected isSameYearAndMonth(date: D, dateToCompare: D): boolean {
```
[the remark this answers is not in the document]
Comment (yggg) [reply src/framework/theme/components/calendar-kit/components/calendar-month-picker/calendar-month-cell.component.ts]: Right, thanks. Done in 3edefa8a95a7ff7ac431c0236578b4e587959e84.

File: src/framework/theme/components/calendar-kit/components/calendar-navigation/calendar-view-mode.component.ts [hunk 19]
```diff
@@ -0,0 +1,72 @@
...
+ * @license
+ * Copyright Akveo. All Rights Reserved.
+ * Licensed under the MIT License. See License.txt in the project root for license information.
+ */
+
+import { ChangeDetectionStrategy, Component, EventEmitter, Input, Output } from '@angular/core';
+import { NbCalendarViewMode } from '../../model';
+import { NbCalendarYearModelService } from '../../services/calendar-year-model.service';
+import { NbDateService } from '../../services/date.service';
+import { TranslationWidth } from '@angular/common';
```
[the remark this answers is not in the document]
Comment (yggg) [reply src/framework/theme/components/calendar-kit/components/calendar-navigation/calendar-view-mode.component.ts]: d29e3cf68c540609b50f194a33134065a26593a6

File: src/framework/theme/components/calendar-kit/components/calendar-picker/calendar-picker.component.ts [hunk 20]
```diff
@@ -35,4 +35,12 @@ export class NbCalendarPickerComponent<D, T> {
   @Input() max: D;
   @Input() filter: (D) => boolean;
   @Output() select: EventEmitter<D> = new EventEmitter();
+
+  @Input()
```
[the remark this answers is not in the document]
Comment (yggg) [reply src/framework/theme/components/calendar-kit/components/calendar-picker/calendar-picker.component.ts]: ¯\_(ツ)_/¯ Fixed by 50ae5607b65bfb4c2f1c760e9183e59898ea3965

File: src/framework/theme/components/calendar-kit/components/calendar-week-number/calendar-week-number.component.ts (hunk 1 above)
Comment (yggg) [reply src/framework/theme/components/calendar-kit/components/calendar-week-number/calendar-week-number.component.ts]: Done in 6d55011424978aede484135f6b688218eea8c9e9. Implemented via `onChanges`. setter, getter require additional field. I believe `ngOnChanges` is more Angular way. What do you think?

File: src/framework/theme/components/calendar-kit/components/calendar-day-picker/calendar-day-cell.component.ts [hunk 21]
```diff
@@ -20,9 +20,12 @@ import { NbDateService } from '../../services/date.service';
 
 @Component({
   selector: 'nb-calendar-day-cell',
-  template: '{{ day }}',
+  template: `
+    <div class="cell-content">
```
[the remark this answers is not in the document]
Comment (yggg) [reply src/framework/theme/components/calendar-kit/components/calendar-day-picker/calendar-day-cell.component.ts]: > Why do we need to add an additional DOM element here? Can't we style this component without a new div element?

We need to combine selection color with different day states. For example, today cell has own background color (white with transparency). It's goes to the `.cell-content`. And we have a blue selection color, which is applied to `.day-cell`, `.month-cell` and `.year-cell`.

File: src/framework/theme/components/calendar-kit/components/calendar-day-picker/calendar-day-cell.component.spec.ts [hunk 22]
```diff
@@ -30,6 +30,7 @@ describe('Component: NbCalendarDayCell', () => {
   });
 
   it('should contain cell class', () => {
+    fixture.detectChanges();
```
[the remark this answers is not in the document]
Comment (yggg) [reply src/framework/theme/components/calendar-kit/components/calendar-day-picker/calendar-day-cell.component.spec.ts]: >  I'm dead sure we need it because of moving class declaration from host param to HostBinding decorator

Right.

> it might be we don't need to move that class declaration to the HostBinding

As discussed above, decorator metadata doesn't inherit. If we keep host class there, styling would break. I believe it's worse than having dumb `cssClass = true` binding.

File: src/framework/theme/components/calendar-kit/components/calendar-month-picker/calendar-month-picker.component.ts [hunk 23]
```diff
@@ -60,18 +73,15 @@ export class NbCalendarMonthPickerComponent<D, T> implements OnInit {
...
+  @HostBinding('class.size-large')
   get large() {
     return this.size === NbCalendarSize.LARGE;
   }
 
-  ngOnInit() {
-    this.initMonths();
+  ngOnChanges(changes: SimpleChanges) {
+    if (changes.month) {
+      this.initMonths();
```
[the remark this answers is not in the document]
Comment (yggg) [reply src/framework/theme/components/calendar-kit/components/calendar-month-picker/calendar-month-picker.component.ts]: What is the benefit of setter over changes hook? It requires additional field, getter.

File: src/framework/theme/components/calendar-kit/components/calendar-navigation/calendar-view-mode.component.spec.ts (hunk 13 above)
[the remark this answers is not in the document]
Comment (yggg) [reply src/framework/theme/components/calendar-kit/components/calendar-navigation/calendar-view-mode.component.spec.ts]: We have `NbCalendarNavigationComponent replaced with NbCalendarViewModeComponent`. As name, inputs, and logic were changed I don't think it worth mention.

File: src/framework/theme/components/calendar-kit/components/calendar-navigation/calendar-view-mode.component.ts [hunk 24]
```diff
@@ -0,0 +1,72 @@
...
+})
+export class NbCalendarViewModeComponent<D> {
+  @Input() date: D;
+  @Input() viewMode: NbCalendarViewMode = NbCalendarViewMode.DATE;
+  @Output() changeMode = new EventEmitter<void>(true);
+
+  constructor(
+    protected dateService: NbDateService<D>,
+    protected yearModelService: NbCalendarYearModelService<D>,
+  ) {}
```
[the remark this answers is not in the document]
Comment (yggg) [reply src/framework/theme/components/calendar-kit/components/calendar-navigation/calendar-view-mode.component.ts]: We don't think we have agreement for constructor parameters formatting. Personally it's my common way for multiple params. It allows to easy change order, add or remove parameter as a line without additional reformatting. I  found 35 occurrences of such formatting across the Nebular.

File: src/framework/theme/components/calendar-kit/components/calendar-week-number/calendar-week-number.component.ts (hunk 2 above)
Comment (yggg) [reply src/framework/theme/components/calendar-kit/components/calendar-week-number/calendar-week-number.component.ts]: > Do we really need an additional container here?

An additional container is required here because the sign cell should be the same height as weekdays row on the right. Weekday row height is a divider (border) width + height of the cell inside. I found, having the sign container with a border + .sign inside simpler and less error-prone than using calc for height.

File: src/framework/theme/components/calendar-kit/components/calendar-year-picker/calendar-year-cell.component.ts (hunk 4 above)
Comment (yggg) [reply src/framework/theme/components/calendar-kit/components/calendar-year-picker/calendar-year-cell.component.ts]: Answered here https://github.com/akveo/nebular/pull/2120#discussion_r355343466

File: src/framework/theme/components/calendar-kit/components/calendar-year-picker/calendar-year-picker.component.ts (hunk 5 above)
Comment (yggg) [reply src/framework/theme/components/calendar-kit/components/calendar-year-picker/calendar-year-picker.component.ts]: See the answer here https://github.com/akveo/nebular/pull/2120#discussion_r355361367.

> I'm pretty sure it fits in one line

It is barely (ends on 117), so I found it more readable to place params on the separate lines.

File: src/framework/theme/components/calendar-kit/services/calendar-year-model.service.spec.ts (hunk 6 above)
Comment (yggg) [reply src/framework/theme/components/calendar-kit/services/calendar-year-model.service.spec.ts]: I tried but can't come up with test cases. Share your ideas and I'll add more tests.

File: src/framework/theme/components/calendar/_calendar.component.theme.scss (hunk 7 above)
Comment (yggg) [reply src/framework/theme/components/calendar/_calendar.component.theme.scss]: `nb-calendar-week-numbers` and `nb-calendar-days-names` borders defined in theme. This line removes ensures it's not shown when there is no navigation bar before. Otherwise, you'll have a double border like this:
image

File: src/framework/theme/components/calendar/calendar-range-cells.ts (hunk 11 above)
Comment (yggg) [reply src/framework/theme/components/calendar/calendar-range-cells.ts]: Done in 29c251ec1c82fc4450ef4cf246936f815eb1550f

File: src/framework/theme/components/calendar-kit/services/calendar-year-model.service.spec.ts (hunk 6 above)
Comment (yggg) [reply src/framework/theme/components/calendar-kit/services/calendar-year-model.service.spec.ts]: Done in 585df46

File: src/framework/theme/components/calendar-kit/components/calendar-day-picker/calendar-day-cell.component.ts (hunk 21 above)
[the remark this answers is not in the document]
Comment (yggg) [reply src/framework/theme/components/calendar-kit/components/calendar-day-picker/calendar-day-cell.component.ts]: Done as discussed irl. 5a09d8c3f
